<html>

<head>
<meta http-equiv="Content-Type" content="text/html; charset=windows-1252">
<meta http-equiv="Content-Language" content="en-us">
<meta name="GENERATOR" content="Microsoft FrontPage 5.0">
<meta name="ProgId" content="FrontPage.Editor.Document">
<title>Sliding MinPD</title>
</head>

<body>

  <center>
  <table border="0" width="729" height="502">
    <tr>
      <td align="center" height="48" width="721"><b><i><font color="#800000" size="6" face="Verdana">Sliding
        MinPD</font></i></b></td>
    </tr>
    <tr>
      <td height="80" width="721">
        <p align="center"><span style="mso-fareast-font-family: Times New Roman; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA"><font color="#000000" size="4" face="Century Gothic"><b>Building
        evolutionary networks of serial samples via a recombination detection
        approach</b></font></span></td>
    </tr>
    <tr>
      <td height="36" width="721"></td>
    </tr>
    <tr>
      <td height="21" width="721">
        <p style="line-height: 100%; margin-top: 0; margin-bottom: 2"><font size="4" face="Verdana"><a href="index.htm">Home</a>
        </font><b><font face="Verdana">|</font></b><font size="4" face="Verdana">&nbsp;
        <a href="Features.htm">Features</a> </font><b><font face="Verdana">|</font></b><font size="4" face="Verdana">&nbsp;
        <a href="Suple.htm">Supplementary Material</a> </font><b><font face="Verdana">
        | </font></b><font size="4" face="Verdana">Source Code</font></p>
      </td>
    </tr>
    <tr>
      <td height="1" width="721" valign="top">
        <p style="line-height: 100%; margin-top: 0; margin-bottom: 0"><b><font face="Verdana">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</font></b></td>
    </tr>
    <tr>
      <td height="28" width="721"></td>
    </tr>
    <tr>
      <td height="27" width="721"><b><font size="4" color="#800000" face="Verdana">Source
        Code</font></b></td>
    </tr>
    <tr>
      <td width="715" height="97" valign="top">
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">The
        C source code for <i>version 1.2 </i>(last updated:<!--webbot bot="Timestamp" S-Type="EDITED" S-Format="%m/%d/%Y" startspan -->06/25/2009<!--webbot bot="Timestamp" i-checksum="12602" endspan -->) 
        is available for download here: </span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">
        <a href="Files/SMinPD1.2.zip">Windows executable and C source code with 
        makefile</a></span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2">
        <span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA; font-weight:700">
        New
        Parameters for version 1.2</span></font></p>
        <font SIZE="2">
        <p><font color="#800000">[markers for clustering] g&quot;markers.txt&quot; [Yes: 
        add g and markers filename, No: g]<br>
        [clustering distance threshold] T0.001<br>
        [clustering option] j4 [0:no clustering, 1:by amino acid, 2:by bases 
        3:only by distance 4:like 1 but post-clustering, 5:like 2 but post, 
        6:like 3 but post]<br>
&nbsp;</font></p>
        </font>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA"> Descriptions of the
        new variables: </span></font></p>
        <ul>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2">M<span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">arkers 
            for Clustering: </span>It is used to cluster similar sequences that 
            share the same amino acids at specific positions (the markers). A 
            sample markers file is included in the tar file above. </font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2">
            Clustering distance threshold: indicates how similar the sequences 
            should be. The default is 0.001 which is equivalent to approx. 1 in 
            1000 nucleotide difference that decides whether two sequences are 
            clustered for similarity (if their distance is less than this 
            threshold)</font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2">
            Clustering Option: Allows the clustering of highly similar sequences 
            before or after distance calculations. For improved recombination 
            detection it is recommended to run pre-clustering (removes 
            sequences), otherwise post-clustering (does not remove sequences).
            </font></li>
        </ul>
        <p style="line-height: 100%"><font face="Verdana" size="2">
        <span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA; font-weight:700">
        Parameters from version 1.0</span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">The parameter file contains all the parameter settings. The
        parameters in <i>Sliding MinPD</i> as they appear in the parameter file 
        (vars.txt) are:&nbsp; </span></font></p>
        <font SIZE="2">
        <p><font color="#800000">[input file] i&quot;in.fas&quot;<br>
        [output file] o&quot;out.txt&quot;<br>
        [report all distances] d0 [0:No, 1:Yes]<br>
        [activate recombination detection] f1 [0:No, 1:Yes]<br>
        [recombination detection option] r2 [1:RIP, 2:B-RIP, 3:SB]<br>
        [crossover option] c1 [0:many, 1:only one]<br>
        [PCC threshold] p0.4<br>
        [window size] w200<br>
        [step size] s20<br>
        [bootstrap recomb. tiebreaker option] t0 [0:No, 1:Yes]<br>
        [bootscan seed] e-3<br>
        [bootscan threshold] h92<br>
        [bootstrap:1/bootknife:0] b0<br>
        [substitution model] vTN93 [options: JC69, K2P, TN93]<br>
        [gamma shape - rate heterogeneity] a0.5 [a0.5]<br>
        [show bootstrap values] k0 [0:No, 1:Yes]</font></p>
        </font>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">The
        program is run with <i><b> &quot;minpd &lt;vars.txt&quot; </b></i> </span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA"> Descriptions of the
        variables: </span></font></p>
        <ul>
          <li>
            <p style="line-height: 100%"><font size="2" face="Verdana"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">Report
            All Distances: writes all pairwise distances between
            ancestor/descendants to a file</span></font></li>
          <li>
            <p style="line-height: 100%"><font size="2" face="Verdana"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">Activate
            Recombination Detection: This option turns the recombination
            detection feature on/off.</span></font></li>
          <li>
            <p style="line-height: 100%"><font size="2" face="Verdana"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">R<font SIZE="2">ecombination
            detection: r2 and r3 run the Bootscan methods that require
            specification of the bootscan seed, bootscan threshold and bootstrap
            or bootknife options values.&nbsp;</font></span></font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">Crossover
            detection: c1 will detect only one crossover, c0 will detect more
            than one.</span></font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">PCC
            threshold: Specifies the Pearson Correlation Coefficient value used
            as a threshold to reduce pool of candidates. A lower value that is
            larger than zero decreases number of false positives.</span></font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">Bootscan
            tiebreaker option: Bootstrap values may add up to more than 100 when
            ancestors with the same distance are present. Use only in
            conjunction with crossover option 1.</span></font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">Bootscan
            Threshold: A threshold used for r2 and r3 to reduce pool of
            candidates.</span></font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">Bootstrap
            or Bootknife: Bootstrap is the standard bootstrap method. Bootfknife
            (as implemented in an earlier version of RDP2) is a hybrid between a
            bootstrap and a jackknife and removes 25% to 50% of sites replacing
            them by other randomly picked sites.&nbsp;</span></font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">Substitution
            model&nbsp; and gamma shape for distance calculations: These are
            required for all recombination detection options. The standard
            bootscan builds NJ trees from a distance matrix and therefore also
            requires specification of these variable.</span></font></li>
          <li>
            <p style="line-height: 100%"><font face="Verdana" size="2">The 
            option to &quot;show bootstrap values&quot; is needed to display the bootstrap 
            values in the network when using the network drawer (see below).
            </font></li>
        </ul>
        <p style="line-height: 100%"><font face="Verdana" size="2">
        <span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA; font-weight:700">
        Input File</span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">The
        <b><i>
        input file</i></b> should be a set of already <b><i>aligned</i></b>
        sequences in <i><b> Fasta</b></i> format with <i><b>sequence IDs </b>
        </i>containing the
        sampling <i><b> time as a prefix</b></i> followed by a dot followed by a sequence
        identifier. The sequences do not need to appear in (sampling) order.
        See sample file &quot;in.fas&quot;</span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2">
        <span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA; font-weight:700">
        Output File</span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">The
        <i><b>
        output</b></i> of Sliding MinPD is a list of the <i><b> ancestor-descendant relationships</b></i>
        and a <i><b> collection of trees </b></i> to build the evolutionary network. 
        </span></font></p>
        <p style="line-height: 100%; margin-bottom:1">
        <font face="Verdana" size="2">The first part contains a list of four 
        semicolon-delimited values:<br>
        Descendant;Ancestor;Distance;Bootstrap support<br>
        <br>
        The recombinant sequences are shown with the notation<br>
        Ancestor=A<sub>1</sub>|BKP<sub>1</sub>|A<sub>2</sub>|BKP<sub>2</sub>|...|BKP<sub>n</sub>|A<sub>n</sub><br>
        Where A<sub>1</sub> is the left ancestral donor, BKP<sub>1</sub> the 
        first breakpoint position, A<sub>2</sub> the ancestral Donor between 
        breakpoints BKP<sub>1</sub> and BKP<sub>2</sub>, etc. A<sub>n</sub> is 
        the right-hand ancestral donor.<br>
        <br>
        <span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA; font-weight:700">
        Network Drawer</span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">The network
        can be build with the online network drawer: <a href="http://72.17.173.2:800/minpd/">http://72.17.173.2:800/minpd/</a> 
        and it</span> is recommended to use smaller scale factors 0.5 instead of 
        1 with larger alignments. The input file is the output file created by 
        MinPD.</font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2">The network 
        drawer outputs the file as a FIG graphic file.
        <a href="http://www.schmidt-web-berlin.de/winfig/">Winfig</a>, a free 
        Windows viewer, can be used to view Fig files. Other viewers can be 
        found here: <a href="http://homepage.usask.ca/~ijm451/fig/">
        http://homepage.usask.ca/~ijm451/fig/</a> </font></p>
        <p style="line-height: 100%"><span style="font-weight: 700">
        <font face="Verdana" size="2">Old</font></span><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA; font-weight:700">
        Version</span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA">
        Version 1.0 is available for download here:&nbsp;</span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA"><a href="Files/SlidingMinPD.zip"> Download C source code
        with makefile</a></span></font></p>
        <p style="line-height: 100%"><font face="Verdana" size="2"><span style="mso-fareast-font-family: SimSun; mso-ansi-language: EN-US; mso-fareast-language: EN-US; mso-bidi-language: AR-SA"><a href="Files/WinSlidingMinPD.zip">Download
        command line (DOS) Windows executable</a></span></font></p>
        <p style="line-height: 100%">&nbsp;</p>
  </table>